base/strings/string.jl

```diff
@@ -245,9 +246,8 @@ function length(s::String, i::Int, j::Int)
         0 ≤ j < ncodeunits(s)+1 || throw(BoundsError(s, j))
     end
     j < i && return 0
-    c = j - i + 1
     @inbounds i, k = thisind(s, i), i
-    c -= i < k
+    c = j - i + (i == k)
```

Comment:
Does this change actually change the behavior? Or is this just an optimization?

Reply:
Actually this is significant. Current (before this PR) behavior is:
```
julia> length("∀", 2, 3)
-1
julia> length("∀∀", 3, 4)
-1
```
this change makes sure that after this PR:
```
julia> length("∀", 2, 3)
0
julia> nlength("∀∀", 3, 4)
1
```

I made this change to avoid rewriting `_length` as you have suggested.

The meaning of this change is that we take `i` *after* aligning it with `thisind` not before it as it was done earlier - this caused the *decrementing* logic in `_length` to decrement too much sometimes (`c` was to small when `i` passed to `length` was inside multy-byte character). This caused a bug as shown in the above examples.

Reply:
Got it. We can merge this as soon as CI passes.